### PR TITLE
core/vm: implement EIP-3541 as an "extra EIP"

### PR DESCRIPTION
**Pull request by chfast**

### CODE REVIEW

File: core/vm/evm.go [hunk 1]
```diff
@@ -463,13 +463,30 @@ func (evm *EVM) create(caller ContractRef, codeAndHash *codeAndHash, gas uint64,
...
+
+	if err == nil {
+		isEIP3541 := func() bool {
+			for _, eip := range evm.vmConfig.ExtraEips {
+				if eip == 3541 {
+					return true
+				}
+			}
+			return false
+		}()
```
Comment (holiman) [inline core/vm/evm.go:479]: I'd just do `evm.chainRules.IsEip3541` -- and later on change it to `IsLondon` or whatever

**Comment by holiman** [comment]
This needs a rebase. I tried doing it myself, but the remote does not allow maintainer push

**Comment by chfast** [comment]
> This needs a rebase. I tried doing it myself, but the remote does not allow maintainer push

This option is not available if a repo is not a personal one.

File: core/vm/evm.go (hunk 1 above)
Comment (chfast) [reply core/vm/evm.go]: I only refactored this. We are happy that we can create tests for `Berlin+3541` for this.

Comment (chfast) [reply core/vm/evm.go]: I only refactored this. We are happy that we can create tests for `Berlin+3541` for this.

Comment (holiman) [reply core/vm/evm.go]: Yeah, that's fine, but in the long run we should have a proper config and tie it to a forkname instead. That ^ is a bit of a hack

File: core/vm/errors.go [hunk 2]
```diff
@@ -34,6 +34,7 @@ var (
 	ErrWriteProtection          = errors.New("write protection")
 	ErrReturnDataOutOfBounds    = errors.New("return data out of bounds")
 	ErrGasUintOverflow          = errors.New("gas uint64 overflow")
+	ErrInvalidCode              = errors.New("invalid code")
```
Comment (axic) [inline core/vm/errors.go:37]: Slightly unrelated question (not blocking this PR): EVMC has a `CONTRACT_VALIDATION_FAILED` error code (used as part of Ewasm 1.0). Probably we could use that error code for EOF1 validation, but should we use it already for this use case, or it makes more sense separating the error codes?

**Comment by chfast** [comment]
Done in #22809.

**Pull request closed without merging**